cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -0,0 +1,804 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "nested_json.hpp"
+
+#include <io/fst/logical_stack.cuh>
+#include <io/fst/lookup_tables.cuh>
+#include <io/utilities/hostdevice_vector.hpp>
+
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/exec_policy.hpp>
+
+namespace cudf::io::json {
+
+//------------------------------------------------------------------------------
+// JSON-TO-STACK-OP DFA
+//------------------------------------------------------------------------------
+namespace to_stack_op {
+
+// Type used to represent the target state in the transition table
+using StateT = char;
+
+/**
+ * @brief Definition of the DFA's states
+ */
+enum class dfa_states : StateT {
+  // The state being active while being outside of a string. When encountering an opening bracket
+  // or curly brace, we push it onto the stack. When encountering a closing bracket or brace, we
+  // pop from the stack.
+  TT_OOS = 0U,
+
+  // The state being active while being within a string (e.g., field name or a string value). We do
+  // not push or pop from the stack while being in this state.
+  TT_STR,
+
+  // The state being active after encountering an escape symbol (e.g., '\'), while being in the
+  // TT_STR state.
+  TT_ESC,
+
+  // Total number of states
+  TT_NUM_STATES
+};
+
+// Aliases for readability of the transition table
+constexpr auto TT_OOS = dfa_states::TT_OOS;
+constexpr auto TT_STR = dfa_states::TT_STR;
+constexpr auto TT_ESC = dfa_states::TT_ESC;
+
+/**
+ * @brief Definition of the symbol groups
+ */
+enum class dfa_symbol_group_id : uint32_t {
```

Comment:
Why do we use `char` for some enum classes but `uint32_t` here? There are certainly less than 256 options, but perhaps there are reasons to desire a 4 byte wide type for thread read alignment?

Reply:
Thanks, I've fixed the type for symbol group ids too. Generally, we prefer to have a smaller type, as we want to "compress" the transition & translation tables. We keep those tables in shared memory and with smaller types we increase the chance of broadcasts and decrease the chance of bank conflicts.